metis-harvesting/src/main/java/eu/europeana/metis/harvesting/http/AbstractHttpHarvestIterator.java

```diff
@@ -0,0 +1,180 @@
+   *
+   * @param action The iteration to perform. It needs to return a result.
+   * @param filter The filter to apply (only records that return true will be sent to the action).
+   * @throws HarvesterException In case there was a problem while harvesting.
+   */
+  public void forEachFileFiltered(ReportingIteration<FullRecord> action, Predicate<Path> filter)
+      throws HarvesterException {
+    forEachPathFiltered(path -> {
+      try (InputStream content = Files.newInputStream(path)) {
+        return action.process(new FullRecordImpl(extractedDirectory.relativize(path).toString(),
+            new ByteArrayInputStream(IOUtils.toByteArray(content))));
+      } catch (RuntimeException e) {
+        throw new IOException("Could not process path " + path + ".", e);
+      }
+    }, filter);
+  }
+
+  @Override
+  public void forEachNonDeleted(ReportingIteration<R> action) throws HarvesterException {
+    forEach(action);
+  }
+
+  @Override
+  public Integer countRecords() throws HarvesterException {
+    // Go by each path only: no need to inspect the full file.
+    final AtomicInteger counter = new AtomicInteger(0);
+    forEachPathFiltered(path -> {
+      counter.incrementAndGet();
+      return IterationResult.CONTINUE;
+    }, path -> true);
+    return counter.get();
+  }
+
+  private static class FileIteration extends SimpleFileVisitor<Path> {
+
+    private static final String MAC_TEMP_FILE = ".DS_Store";
+    private static final String MAC_TEMP_FOLDER = "__MACOSX";
+
+    private final ReportingIteration<Path> action;
+    private final Predicate<Path> filter;
+
+    public FileIteration(ReportingIteration<Path> action, Predicate<Path> filter) {
+      this.action = action;
+      this.filter = filter;
+    }
+
+    @Override
+    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+      if (!filter.test(file)) {
+        return FileVisitResult.CONTINUE;
+      }
+      final Path fileName = file.getFileName();
+      if (fileName != null && MAC_TEMP_FILE.equals(fileName.toString())) {
+        return FileVisitResult.CONTINUE;
+      }
+      if (CompressedFileExtension.forPath(file) != null) {
+        return FileVisitResult.CONTINUE;
+      }
+      final IterationResult result = action.process(file);
+      if (result == null) {
+        throw new IllegalArgumentException("Iteration result cannot be null.");
+      }
+      return result == IterationResult.TERMINATE ? FileVisitResult.TERMINATE
+          : FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+      final Path dirName = dir.getFileName();
+      if (dirName != null && MAC_TEMP_FOLDER.equals(dirName.toString())) {
+        return FileVisitResult.SKIP_SUBTREE;
+      }
+      return FileVisitResult.CONTINUE;
+    }
+  }
+
+  record FullRecordImpl(String relativeFilePath, ByteArrayInputStream entryContent) implements FullRecord {
```

Comment:
should be this also `private` as before the class extraction?

Reply:
yes I think it's better indeed